README.md
# selectable_container

A Selectable Container for Flutter. Easy to implement and customize.

## Demo
![Demo: ](https://github.com/fgcan10/selectable_container/raw/master/selectableContainer.gif)

## Usage
To use plugin, just import package `import 'package:selectable_container/selectable_container.dart';`

## Example
Check de example directory to access the project that implements the image shown above.

````dart
    import 'package:flutter/material.dart';
    import 'package:selectable_container/selectable_container.dart';

    class SimpleExamplePage extends StatefulWidget {
      @override
      _SimpleExamplePageState createState() => _SimpleExamplePageState();
    }

    class _SimpleExamplePageState extends State<SimpleExamplePage> {
      @override
      Widget build(BuildContext context) {
        return Scaffold(
          appBar: AppBar(
            title: Text('Selectable Container'),
          ),
          body: Column(
            children: <Widget>[
              Text(
                'Default container',
                style: TextStyle(fontSize: 22, fontWeight: FontWeight.bold),
              ),
              SelectableContainer(
                selected: _selected
                onValueChanged: (newValue) {
                   setState(() {
                    _selected = newValue;
                  });
                },
                child: Text('Tap me'),
                padding: 16.0,
              ),
            ],
          ),
        );
      }
    }
````


CHANGELOG.md
## [0.0.1] - 2019-11-10

* A selectable container for Flutter. Easy to implement and customize.

## [0.0.2] - 2019-11-10

* Beter license and documentation

## [0.1.1] - 2019-11-10

* All settings tested. Production version.

## [0.1.2] - 2019-11-15

* New setting 'elevation' and 'borderradius'

## [0.1.3] - 2019-11-28

* Property selected is now public

## [0.1.4] - 2020-07-20

* Property selected is now public

## [0.1.5] - 2021-03-21

* onPressed Deprecated. onValueChanged added.
* Widget is now stateless.

## [0.1.6] - 2021-03-25

* Display a different icon when unselected

## [1.0.1] - 2021-04-03

* Null-safety

## [1.0.2] - 2021-07-14

* Select/unselect icon

## [1.0.3] - 2022-06-18

* Top, botton, left and right margin

## [1.0.4] - 2022-07-02

* New properties : selectedBackgroundColorIcon, unselectedBackgroundColorIcon selectedBorderColorIcon and unselectedBorderColorIcon

## [1.0.5] - 2022-07-02

* New properties : topIconPosition, bottomIconPosition, leftIconPosition, rightIconPosition

## [1.0.6] - 2022-07-02

* No border (borderSize=0) bug solved.

## [1.0.7] - 2022-27-10

* Bug related with transparent margin solved. Thanks to rvndsngwn

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:example2/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


test/selectable_container_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:selectable_container/selectable_container.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


example/lib/main.dart
import 'package:example2/selectable_container_grid.dart';
import 'package:flutter/material.dart';
import 'package:selectable_container/selectable_container.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Selectable Container Example',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  bool _select1 = false;
  bool _select2 = false;
  bool _select3 = false;

  @override
  Widget build(BuildContext context) {
    var textStyles = Theme.of(context).textTheme;

    return Scaffold(
      backgroundColor: Colors.red,
      appBar: AppBar(
        title: const Text('Selectable Container Example'),
      ),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.start,
        children: [
          Align(
            alignment: Alignment.bottomCenter,
            child: SelectableContainer(
              onValueChanged: (newValue) {
                setState(() {
                  _select1 = newValue;
                });
              },
              marginColor: Colors.transparent,
              selected: _select1,
              padding: 8.0,
              child: buildTextContentOfContainer(
                  'Selectable Container', 'Default theme colors', textStyles),
            ),
          ),
          const SizedBox(height: 16.0),
          SelectableContainer(
            selectedBorderColor: Colors.teal.shade700,
            selectedBackgroundColor: Colors.grey.shade100,
            unselectedBorderColor: Colors.teal.shade600,
            unselectedBackgroundColor: Colors.grey.shade200,
            iconAlignment: Alignment.topLeft,
            icon: Icons.thumb_up,
            unselectedIcon: Icons.add_box,
            marginColor: Colors.transparent,
            elevation: 0,
            iconSize: 24,
            unselectedOpacity: 1,
            selectedOpacity: 0.8,
            selected: _select2,
            padding: 8.0,
            child: buildTextContentOfContainer('Custom color',
                'Icon, size, position en opacitiy changed', textStyles),
            onValueChanged: (newValue) {
              setState(() {
                _select2 = newValue;
              });
            },
          ),
          const SizedBox(height: 16.0),
          SelectableContainer(
            iconSize: 20,
            icon: Icons.star,
            iconColor: Colors.yellowAccent,
            marginColor: Colors.transparent,
            unselectedOpacity: 0.3,
            opacityAnimationDuration: 300,
            elevation: 4.0,
            selected: _select3,
            padding: 16.0,
            onValueChanged: (newValue) {
              setState(() {
                _select3 = newValue;
              });
            },
            child: buildDemoContent3(),
          ),
          const SizedBox(height: 16.0),
          const Expanded(child: SelectableContainerGrid()),
        ],
      ),
    );
  }

  Widget buildTextContentOfContainer(
      String title, String subtitle, TextTheme textStyles) {
    return Column(
      mainAxisSize: MainAxisSize.min,
      crossAxisAlignment: CrossAxisAlignment.center,
      children: <Widget>[
        Text(
          title,
          style: textStyles.headline5,
        ),
        Text(
          subtitle,
          style: textStyles.bodyText1,
        ),
      ],
    );
  }

  Widget buildDemoContent3() {
    return Column(
      children: <Widget>[
        Row(
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            const CircleAvatar(
              backgroundImage: AssetImage('images/flutter.png'),
              radius: 25.0,
            ),
            const SizedBox(
              width: 8.0,
            ),
            Text(
              'Develop with Flutter',
              style: TextStyle(
                  fontWeight: FontWeight.w500,
                  fontSize: 20.0,
                  color: Colors.blue.shade700),
            ),
          ],
        )
      ],
    );
  }
}


lib/selectable_container.dart
library selectable_container;

import 'package:flutter/material.dart';

/// Box that can be tapped.
/// When selected a check Icon appears
class SelectableContainer extends StatelessWidget {
  /// Background color for the space between the child and the
  /// Defaul value: scaffoldBackgroundColor
  final Color? marginColor;

  /// Background color when container is selected.
  /// Default value : dialogBackgroundColor
  final Color? selectedBackgroundColor;

  /// Background color when container is not selected.
  /// Default value : dialogBackgroundColor
  final Color? unselectedBackgroundColor;

  /// Background color of the icon when container is selected.
  /// Default value : theme.primaryColor
  final Color? selectedBackgroundColorIcon;

  /// Background color of the icon when container is not selected.
  /// Default value : theme.primaryColorDark
  final Color? unselectedBackgroundColorIcon;

  /// Border color when container is selected.
  /// Default value : primaryColor
  final Color? selectedBorderColor;

  /// Border color when container is not selected.
  /// Default value :primaryColorDark
  final Color? unselectedBorderColor;

  /// Border color of the icon when container is selected.
  /// Default value : Colors.white
  final Color? selectedBorderColorIcon;

  /// Border color of the icon when container is not selected.
  /// Default value : Colors.white
  final Color? unselectedBorderColorIcon;

  /// Icon's color
  /// Default value : white
  final Color iconColor;

  /// Icon's size
  /// Default value : 16
  final int iconSize;

  /// The child to render inside the container
  final Widget? child;

  /// Border size
  /// Default value : 2 pixels
  final int borderSize;

  /// Callback when container get tapped
  final Function? onPressed;

  /// Callback of type ValueChanged
  final ValueChanged<bool> onValueChanged;

  /// Opacity when container is not selected.
  /// When not 1 it will be animated when tapped.
  /// Default value : 0.5
  final double unselectedOpacity;

  /// Opacity when container is selected.
  /// When not 1 it will be animated when tapped.
  /// Default value : 1
  final double selectedOpacity;

  /// Opacity animation duration in milliseconds.
  /// Default value : 600
  final int opacityAnimationDuration;

  /// Icon to be shown when selected.
  /// Default value : Icons.check
  final IconData icon;

  /// Icon to be shown when unselected.
  /// Default value : geen
  final IconData? unselectedIcon;

  /// Icon position.
  /// Default value : Alignment.topRight
  final Alignment iconAlignment;

  /// Padding of child content
  /// Default value : 0.0
  final double padding;

  /// Elevation
  /// Default value : 0.0
  final double elevation;

  ///Default value : 10.0
  final double borderRadius;

  ///Default not selected
  final bool selected;

  // Top marging. Default value : 0.0
  final double topMargin;

  // Bottom marging. Default value : 0.0
  final double bottomMargin;

  // Left marging. Default value : 0.0
  final double leftMargin;

  // Right marging. Default value : 0.0
  final double rightMargin;

  // Top position of the icon.
  final double? topIconPosition;

  // Bottom position of the icon.
  final double? bottomIconPosition;

  // Left position of the icon.
  final double? leftIconPosition;

  // Right position of the icon.
  final double? rightIconPosition;

  SelectableContainer(
      {required this.selected,
      this.marginColor,
      this.unselectedBackgroundColor,
      this.selectedBackgroundColor,
      this.unselectedBackgroundColorIcon,
      this.selectedBackgroundColorIcon,
      this.selectedBorderColor,
      this.unselectedBorderColor,
      this.selectedBorderColorIcon,
      this.unselectedBorderColorIcon,
      @Deprecated('Use onValueChanged') this.onPressed,
      required this.onValueChanged,
      this.iconSize = 16,
      this.iconColor = Colors.white,
      this.icon = Icons.check,
      this.unselectedIcon,
      this.iconAlignment = Alignment.topRight,
      this.borderSize = 2,
      this.selectedOpacity = 1.0,
      this.unselectedOpacity = 0.5,
      this.opacityAnimationDuration = 600,
      this.padding = 0,
      this.elevation = 0.0,
      this.borderRadius = 10.0,
      this.topMargin = 0.0,
      this.bottomMargin = 0.0,
      this.leftMargin = 0.0,
      this.rightMargin = 0.0,
      this.topIconPosition,
      this.bottomIconPosition,
      this.leftIconPosition,
      this.rightIconPosition,
      required this.child});

  @override
  Widget build(BuildContext context) {
    var theme = Theme.of(context);
    return GestureDetector(
      onTap: () {
        onValueChanged(!selected);
      },
      child: AnimatedOpacity(
        opacity: selected ? selectedOpacity : unselectedOpacity,
        duration: Duration(milliseconds: opacityAnimationDuration),
        child: Material(
          elevation: elevation,
          color: marginColor ?? theme.scaffoldBackgroundColor,
          child: Padding(
            padding: EdgeInsets.fromLTRB(
                leftMargin, topMargin, rightMargin, bottomMargin),
            child: Stack(
              alignment: iconAlignment,
              children: <Widget>[
                Container(
                  clipBehavior: Clip.antiAliasWithSaveLayer,
                  margin: EdgeInsets.all(iconSize / 2),
                  padding: EdgeInsets.all(padding),
                  child: child,
                  decoration: BoxDecoration(
                      border: borderSize > 0
                          ? Border.all(
                              color: selected
                                  ? selectedBorderColor ?? theme.primaryColor
                                  : unselectedBorderColor ??
                                      theme.primaryColorDark,
                              width: borderSize.toDouble(),
                            )
                          : Border.symmetric(),
                      borderRadius: BorderRadius.circular(borderRadius),
                      color: selected
                          ? selectedBackgroundColor ??
                              theme.dialogBackgroundColor
                          : unselectedBackgroundColor ??
                              theme.dialogBackgroundColor),
                ),
                Positioned(
                  top: topIconPosition,
                  bottom: bottomIconPosition,
                  left: leftIconPosition,
                  right: rightIconPosition,
                  child: Visibility(
                      visible: !selected && unselectedIcon != null,
                      child: Container(
                        padding: EdgeInsets.all(4),
                        decoration: BoxDecoration(
                            border: Border.all(
                                color:
                                    unselectedBorderColorIcon ?? Colors.white),
                            shape: BoxShape.circle,
                            color: unselectedBackgroundColorIcon ??
                                theme.primaryColorDark),
                        child: Icon(
                          unselectedIcon,
                          size: iconSize.toDouble(),
                          color: iconColor,
                        ),
                      )),
                ),
                Positioned(
                  top: topIconPosition,
                  bottom: bottomIconPosition,
                  left: leftIconPosition,
                  right: rightIconPosition,
                  child: Visibility(
                    visible: selected,
                    child: Container(
                      padding: EdgeInsets.all(4),
                      decoration: BoxDecoration(
                          border: Border.all(
                              color: selectedBorderColorIcon ?? Colors.white),
                          shape: BoxShape.circle,
                          color: selectedBackgroundColorIcon ??
                              theme.primaryColor),
                      child: Icon(
                        icon,
                        size: iconSize.toDouble(),
                        color: iconColor,
                      ),
                    ),
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}


example/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.CreateAndShow(L"example2", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


example/lib/selectable_container_grid.dart
import 'package:flutter/material.dart';
import 'package:selectable_container/selectable_container.dart';

final List<Color> kGridElements = [
  Colors.grey[50]!,
  Colors.grey[200]!,
  Colors.grey[400]!,
  Colors.brown[100]!,
  Colors.brown[200]!,
  Colors.brown[400]!,
];

class SelectableContainerGrid extends StatefulWidget {
  const SelectableContainerGrid({Key? key}) : super(key: key);

  @override
  // ignore: library_private_types_in_public_api
  _SelectableContainerGridState createState() =>
      _SelectableContainerGridState();
}

class _SelectableContainerGridState extends State<SelectableContainerGrid> {
  final List<bool> _selectedItems = [];

  @override
  void initState() {
    super.initState();
    for (var element in kGridElements) {
      _selectedItems.add(false);
    }
  }

  @override
  Widget build(BuildContext context) {
    return GridView.count(
      crossAxisCount: 3,
      children: _buildElements(),
    );
  }

  List<Widget> _buildElements() {
    List<Widget> ret = [];

    for (var i = 0; i < kGridElements.length; i++) {
      ret.add(SelectableContainer(
        selectedBackgroundColor: kGridElements[i],
        unselectedBackgroundColor: kGridElements[i],
        borderSize: 0,
        selectedBackgroundColorIcon: Theme.of(context).dialogBackgroundColor,
        unselectedBackgroundColorIcon: Theme.of(context).dialogBackgroundColor,
        selected: _selectedItems[i],
        unselectedIcon: Icons.check_box_outline_blank,
        icon: Icons.check_box,
        iconAlignment: Alignment.topRight,
        iconSize: 24,
        iconColor: Colors.deepOrange[400]!,
        onValueChanged: (_) {
          _valueSelected(i);
        },
        topMargin: 5.0,
        leftMargin: 10.0,
        rightMargin: 10.0,
        bottomMargin: 10.0,
        child: SizedBox(
          height: 100,
          width: 100,
          child: Image.asset(
            'images/dog$i.jpg',
            fit: BoxFit.fitWidth,
          ),
        ),
      ));
    }
    return ret;
  }

  void _valueSelected(int index) {
    setState(() {
      _selectedItems[index] = !_selectedItems[index];
    });
  }
}


example/lib/simpleExample.dart
import 'package:flutter/material.dart';
import 'package:selectable_container/selectable_container.dart';

class SimpleExamplePage extends StatefulWidget {
  @override
  _SimpleExamplePageState createState() => _SimpleExamplePageState();
}

class _SimpleExamplePageState extends State<SimpleExamplePage> {
  bool _selected = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Selectable Container'),
      ),
      body: Column(
        children: <Widget>[
          Text(
            'Default container',
            style: TextStyle(fontSize: 22, fontWeight: FontWeight.bold),
          ),
          SelectableContainer(
            selected: _selected,
            onValueChanged: (newValue) {
              setState(() {
                _selected = newValue;
              });
            },
            child: Text('Tap me'),
            padding: 16.0,
          ),
        ],
      ),
    );
  }
}
